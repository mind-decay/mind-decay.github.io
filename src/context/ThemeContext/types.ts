import { Nullable } from 'types';

export type ThemeType = 'light' | 'dark';

export type ThemeContextType = {
  theme: ThemeType;
  toggleTheme: () => void;
};

export const isThemeType = (theme: Nullable<string>): theme is ThemeType => {
  return typeof theme === 'string' && ['light', 'dark'].includes(theme);
};
