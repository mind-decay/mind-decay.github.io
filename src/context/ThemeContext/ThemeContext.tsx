import { ReactNode, createContext, useContext, useEffect, useState } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';

import { ThemeContextType, ThemeType, isThemeType } from './types';
import { Nullable } from 'types/helpers';

import { darkTheme, lightTheme } from 'styles/themes';

const ThemeContext = createContext<Nullable<ThemeContextType>>(null);

export const useTheme = (): ThemeContextType => {
  const context = useContext(ThemeContext);

  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }

  return context;
};

type ThemeProviderProps = {
  children: ReactNode;
};

export const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [theme, setTheme] = useState<ThemeType>('dark');
  const isLightTheme = theme === 'light';

  const toggleTheme = () => {
    const updatedTheme = isLightTheme ? 'dark' : 'light';
    setTheme(updatedTheme);
    localStorage.setItem('theme', updatedTheme);
  };

  useEffect(() => {
    const savedTheme = localStorage.getItem('theme');
    const prefersLight = window.matchMedia && window.matchMedia('(prefers-color-scheme: light)').matches;
    const checkCurrentTheme = isThemeType(savedTheme);

    if (checkCurrentTheme) {
      setTheme(savedTheme);
    } else if (prefersLight) {
      setTheme('light');
    }
  }, []);

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      <StyledThemeProvider theme={isLightTheme ? lightTheme : darkTheme}>{children}</StyledThemeProvider>
    </ThemeContext.Provider>
  );
};
