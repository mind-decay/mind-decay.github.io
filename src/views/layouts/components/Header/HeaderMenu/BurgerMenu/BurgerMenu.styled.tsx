import styled from 'styled-components';

export const StyledBurgerMenu = styled.button`
  order: 1;

  svg {
    display: flex;
    width: 30px;
    height: 30px;
    cursor: pointer;
    transform: scaleX(-1);
    line {
      stroke: ${({ theme }) => theme.colors.primary};
      stroke-dasharray: 100%;
      stroke-dashoffset: 0%;
      transition:
        transform 0.3s,
        stroke-dashoffset 0.4s;
      transform-origin: center;
      stroke-width: 3px;

      &:first-child {
        transform: translateY(-35%);
      }

      &:nth-child(2) {
        transform: scaleX(1);
        transition:
          opacity 0.3s,
          transform 0.3s;
      }

      &:last-child {
        transform: translateY(33%);
        stroke-dasharray: 100%;
        stroke-dashoffset: 9px;
      }
    }
  }

  &:hover {
    svg {
      line {
        &:last-child {
          stroke-dashoffset: 0%;
        }
      }
    }
  }

  &.active {
    svg {
      line {
        &:first-child {
          transform: rotate(45deg);
          stroke-linecap: round;
        }

        &:nth-child(2) {
          transform: scaleX(0.1);
          opacity: 0;
        }

        &:last-child {
          transform: rotate(-45deg);
          stroke-linecap: round;
          stroke-dashoffset: 0%;
        }
      }
    }
  }
`;
