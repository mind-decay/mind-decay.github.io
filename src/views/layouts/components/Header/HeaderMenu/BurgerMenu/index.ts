export * from './BurgerMenu';
export * from './BurgerMenu.styled';
