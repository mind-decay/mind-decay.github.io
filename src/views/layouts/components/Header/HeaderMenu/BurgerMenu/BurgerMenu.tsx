import { StyledBurgerMenu } from './BurgerMenu.styled';

type BurgerMenuProps = {
  isVisible: boolean;
  onClick: () => void;
};

export const BurgerMenu = ({ isVisible, onClick }: BurgerMenuProps) => {
  return (
    <StyledBurgerMenu
      type="button"
      onClick={onClick}
      className={`${isVisible ? 'active' : ''}`}
      aria-label="Открыть меню"
    >
      <svg>
        <line x1="0" y1="50%" x2="100%" y2="50%" shapeRendering="crispEdges" />
        <line x1="0" y1="50%" x2="100%" y2="50%" shapeRendering="crispEdges" />
        <line x1="0" y1="50%" x2="100%" y2="50%" shapeRendering="crispEdges" />
      </svg>
    </StyledBurgerMenu>
  );
};
