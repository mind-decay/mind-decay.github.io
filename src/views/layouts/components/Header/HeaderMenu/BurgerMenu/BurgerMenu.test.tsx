import { vi } from 'vitest';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import { ThemeProvider } from 'context/ThemeContext';

import { BurgerMenu } from './BurgerMenu';

test('renders burger menu', () => {
  const handler = vi.fn();
  const { container } = render(
    <ThemeProvider>
      <BurgerMenu isVisible={true} onClick={handler} />
    </ThemeProvider>,
  );

  expect(container).toMatchSnapshot();
});
