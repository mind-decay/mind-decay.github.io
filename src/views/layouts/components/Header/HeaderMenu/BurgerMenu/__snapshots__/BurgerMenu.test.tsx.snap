// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders burger menu 1`] = `
<div>
  <button
    aria-label="Открыть меню"
    class="sc-beySPh ihpJIx active"
    type="button"
  >
    <svg>
      <line
        shape-rendering="crispEdges"
        x1="0"
        x2="100%"
        y1="50%"
        y2="50%"
      />
      <line
        shape-rendering="crispEdges"
        x1="0"
        x2="100%"
        y1="50%"
        y2="50%"
      />
      <line
        shape-rendering="crispEdges"
        x1="0"
        x2="100%"
        y1="50%"
        y2="50%"
      />
    </svg>
  </button>
</div>
`;
